solarkraft/src/fetcher/callDecoder.ts

```diff
@@ -0,0 +1,178 @@
+/**
+ * @license
+ * [Apache-2.0](https://github.com/freespek/solarkraft/blob/main/LICENSE)
+ */
+
+/**
+ * The code to extract Stellar operations that correspond to Soroban contract calls.
+ *
+ * Igor Konnov, 2024.
+ */
+
+import sdk, { Address } from '@stellar/stellar-sdk'
+import { ContractCallEntry, FieldsMap } from './storage.js'
+import { Maybe, just, none } from '@sweet-monads/maybe'
+import { OrderedMap } from 'immutable'
+
+/**
+ * Decode a contract call from a Horizon operation.
+ * @param op an operation
+ * @param matcher a quick matcher over the contractId to avoid expensive deserialization
+ */
+export async function extractContractCall(
+    op: any,
+    matcher: (contractId: string) => boolean
+): Promise<Maybe<ContractCallEntry>> {
+    // https://developers.stellar.org/network/horizon/api-reference/resources/operations/object
+    if (op.function === 'HostFunctionTypeHostFunctionTypeInvokeContract') {
+        // This operation represents a smart contract call.
+        // In particular, it contains: the callee contract, the method, and its parameters.
+        //
+        // https://developers.stellar.org/network/horizon/api-reference/resources/operations/object/invoke-host-function
+
+        // extract the call data: transaction hash, parameters
+        const txHash = op.transaction_hash
+        // decode the call parameters from XDR to native JS values
+        const params = op.parameters.map((p) => {
+            return sdk.scValToNative(sdk.xdr.ScVal.fromXDR(p.value, 'base64'))
+        })
+        const contractId = params[0]
+        if (!matcher(contractId)) {
+            // return quickly to avoid additional calls and deserialization
+            return none()
+        }
+
+        // continue with extracting transaction data and ledger updates
+        const method = params[1]
+        const methodArgs = params.slice(2)
+
+        const tx = await op.transaction()
+        // is it the right way to extract the height?
```

Comment:
Is this a TODO?

Reply:
I mean, yeah. I hope someone more knowledgeable would read it tells me to fix it :)

Reply:
added `TODO:` for the future